### ISSUE TITLE
Downgrade of key in PKCS7_SIGNER_INFO_set()

### ISSUE THREAD

**Issue by mattcaswell** [openssl/openssl#12991]
We are deliberately downgrading a key in PKCS7_SIGNER_INFO_set() which suggests missing functionality. Do we need to do something?

Reference:

https://github.com/openssl/openssl/blob/a48309cb5c58be8fa310608124925d4729664cf8/crypto/pkcs7/pk7_lib.c#L316-L320

**Comment by slontis** [comment]
There are multiple ameth ctrls that are used in pkcs7
Not long after the downgrade it calls 
ret = pkey->ameth->pkey_ctrl(pkey, ASN1_PKEY_CTRL_PKCS7_SIGN, 0, p7i);
Which has different code for RSA, DSA and EC?

I assume we dont really want this code sitting in a provider?

````
    case ASN1_PKEY_CTRL_PKCS7_SIGN:
        if (arg1 == 0) {
            int snid, hnid;
            X509_ALGOR *alg1, *alg2;

            PKCS7_SIGNER_INFO_get0_algs(arg2, NULL, &alg1, &alg2);
            if (alg1 == NULL || alg1->algorithm == NULL)
                return -1;
            hnid = OBJ_obj2nid(alg1->algorithm);
            if (hnid == NID_undef)
                return -1;
            if (!OBJ_find_sigid_by_algs(&snid, hnid, EVP_PKEY_id(pkey)))
                return -1;
            X509_ALGOR_set0(alg2, OBJ_nid2obj(snid), V_ASN1_UNDEF, 0);
        }
````
 NOTE: there is also a 
ret = pkey->ameth->pkey_ctrl(pkey, ASN1_PKEY_CTRL_PKCS7_ENCRYPT, 0, p7i);

**Comment by levitte** [comment]
You only have to look at this line:

``` C
            PKCS7_SIGNER_INFO_get0_algs(arg2, NULL, &alg1, &alg2);
```

.... and look up the function signature:

``` console
$ grep -A1 PKCS7_SIGNER_INFO_get0_algs /usr/include/openssl/pkcs7.h 
void PKCS7_SIGNER_INFO_get0_algs(PKCS7_SIGNER_INFO *si, EVP_PKEY **pk,
                                 X509_ALGOR **pdig, X509_ALGOR **psig);
```

.... to see that these controls relies on libcrypto structures being passed all the way deep into the backend.  This does not translate to providers, at all.

Similar stuff has already been worked on.  Have a look at `crypto/asn1/a_sign.c`, `ASN1_item_sign_ctx()`, there's provider specific stuff there that more or less does what the ASN1_PKEY_CTRL_PKCS7_SIGN shown above does.

The ASN1_PKEY_CTRL_PKCS7_SIGN shown above is easy enough, really.  All it does is extract an algorithm IDs, which has been solved (see `ASN1_item_sign_ctx()`, as mentioned already), even though similar coding may be required in the pkcs7 sub-library.

All the engine implementations I've seen so far are gentle, at least with this specific ctrl command.  They all seem to only reach for algorithm identities and that's about it, with the OSSL_PARAM code from `ASN1_item_sign_ctx()` for guidance.

Please please please make new OSSL_PARAM names for this kind of stuff fairly generic, so we don't end up with a gazillion special cases.

**Comment by mattcaswell** [comment]
> .... to see that these controls relies on libcrypto structures being passed all the way deep into the backend. This does not translate to providers, at all.

This looks like much the same issue as was solved by #13088 for the CMS code.

**Comment by levitte** [comment]
> This looks like much the same issue as was solved by #13088 for the CMS code.

Yup.  I did point that out there.  You response was...  interesting :wink:

https://github.com/openssl/openssl/pull/13088#issuecomment-706020725

**Comment by slontis** [comment]
I will have a crack at it.. Share the pain I say.. :)
